src/boot/efi/boot.c

```diff
@@ -2222,6 +2270,125 @@ static void config_entry_add_linux(
         }
 }
 
+static EFI_STATUS secure_boot_enroll(EFI_FILE *root_dir, Config *config, ConfigEntry *entry) {
+        assert(root_dir);
+        assert(config);
+        assert(entry);
+
+        typedef struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+                CHAR8 *buffer;
+                UINTN size;
+        } SecureBootVariable;
+
+        static SecureBootVariable sb_vars[] = {
+                { L"db",  L"db.esl",  EFI_IMAGE_SECURITY_DATABASE_VARIABLE, NULL, 0 },
+                { L"KEK", L"KEK.esl", EFI_GLOBAL_VARIABLE, NULL, 0 },
+                { L"PK",  L"PK.esl",  EFI_GLOBAL_VARIABLE, NULL, 0 },
+        };
+
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_freepool_ CHAR16 *keys_dir = NULL;
+        _cleanup_(file_closep) EFI_FILE_HANDLE dir = NULL;
+        UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS;
+
+        keys_dir = PoolPrint(L"\\loader\\keys\\%s", entry->current_name);
+        if (!keys_dir)
+                return log_oom();
+
+        err = open_directory(root_dir, keys_dir, &dir);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed opening keys directory %s: %r", keys_dir, err);
+
+        /* we start by loading ALL the variables from the disk */
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                err = file_read(dir, sb_vars[i].filename, 0, 0, &sb_vars[i].buffer, &sb_vars[i].size);
+                if (EFI_ERROR(err)) {
+                        log_error_stall(L"Failed reading file %s: %r", sb_vars[i].filename, err);
+                        goto out_deallocate;
+                }
+        }
+
+        /* we then write them to the efivars */
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, sb_vars[i].buffer, sb_vars[i].size, sb_vars_opts);
+                if (EFI_ERROR(err)) {
+                        log_error_stall(L"Failed to write %s secure boot variable: %r", sb_vars[i].name, err);
+                        goto out_deallocate;
+                }
+        }
+
+        /* if we loaded the signature keys successfully then we remove all the
+         * signature keys entries from  the menu as the system is now locked down. */
+        config_remove_by_type(config, LOADER_SIGNATURE_KEYS);
```

Comment:
To be honest, I don't think this complication is really necessary. I would rather we unconditionally reboot after enrolling keys and not bother trying to clean up the UI. In the unlikely case where reboot fails, it's a "so what", next boot these entries will vanish anyways.

There is even one good reason to reboot anyway: TPM PCR 7 measurements (possibly others) should be changing with this reboot (after all, we *just* changed secure boot state) and we want to be nice to the loaded OS. If it wants to use it for encryption keys enrollment right away, than it would find that it cannot unseal its keys on next boot. Rebooting unconditionally ensures the TPM is in an expected state.

Reply:
That was part of my initial patch, I do agree that it would probably be the cleanest way to dot it.